doc/man1/momctl.1.in
.\"         OpenPBS (Portable Batch System) v2.3 Software License
.\" 
.\" Copyright (c) 1999-2000 Veridian Information Solutions, Inc.
.\" All rights reserved.
.\" 
.\" ---------------------------------------------------------------------------
.\" For a license to use or redistribute the OpenPBS software under conditions
.\" other than those described below, or to purchase support for this software,
.\" please contact Veridian Systems, PBS Products Department ("Licensor") at:
.\" 
.\"    www.OpenPBS.org  +1 650 967-4675                  sales@OpenPBS.org
.\"                        877 902-4PBS (US toll-free)
.\" ---------------------------------------------------------------------------
.\" 
.\" This license covers use of the OpenPBS v2.3 software (the "Software") at
.\" your site or location, and, for certain users, redistribution of the
.\" Software to other sites and locations.  Use and redistribution of
.\" OpenPBS v2.3 in source and binary forms, with or without modification,
.\" are permitted provided that all of the following conditions are met.
.\" After December 31, 2001, only conditions 3-6 must be met:
.\" 
.\" 1. Commercial and/or non-commercial use of the Software is permitted
.\"    provided a current software registration is on file at www.OpenPBS.org.
.\"    If use of this software contributes to a publication, product, or service
.\"    proper attribution must be given; see www.OpenPBS.org/credit.html
.\" 
.\" 2. Redistribution in any form is only permitted for non-commercial,
.\"    non-profit purposes.  There can be no charge for the Software or any
.\"    software incorporating the Software.  Further, there can be no
.\"    expectation of revenue generated as a consequence of redistributing
.\"    the Software.
.\" 
.\" 3. Any Redistribution of source code must retain the above copyright notice
.\"    and the acknowledgment contained in paragraph 6, this list of conditions
.\"    and the disclaimer contained in paragraph 7.
.\" 
.\" 4. Any Redistribution in binary form must reproduce the above copyright
.\"    notice and the acknowledgment contained in paragraph 6, this list of
.\"    conditions and the disclaimer contained in paragraph 7 in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" 5. Redistributions in any form must be accompanied by information on how to
.\"    obtain complete source code for the OpenPBS software and any
.\"    modifications and/or additions to the OpenPBS software.  The source code
.\"    must either be included in the distribution or be available for no more
.\"    than the cost of distribution plus a nominal fee, and all modifications
.\"    and additions to the Software must be freely redistributable by any party
.\"    (including Licensor) without restriction.
.\" 
.\" 6. All advertising materials mentioning features or use of the Software must
.\"    display the following acknowledgment:
.\" 
.\"     "This product includes software developed by NASA Ames Research Center,
.\"     Lawrence Livermore National Laboratory, and Veridian Information
.\"     Solutions, Inc.
.\"     Visit www.OpenPBS.org for OpenPBS software support,
.\"     products, and information."
.\" 
.\" 7. DISCLAIMER OF WARRANTY
.\" 
.\" THIS SOFTWARE IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND. ANY EXPRESS
.\" OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT
.\" ARE EXPRESSLY DISCLAIMED.
.\" 
.\" IN NO EVENT SHALL VERIDIAN CORPORATION, ITS AFFILIATED COMPANIES, OR THE
.\" U.S. GOVERNMENT OR ANY OF ITS AGENCIES BE LIABLE FOR ANY DIRECT OR INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
.\" OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
.\" EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" 
.\" This license will be governed by the laws of the Commonwealth of Virginia,
.\" without reference to its choice of law rules.
.if \n(Pb .ig Iq
.TH momctl 1B "" Local PBS
.so ../ers/ers.macros
.Iq
.SH NAME
qdel \- Remote shutdown, reconfiguration, diagnostics, and querying of the pbs_mom daemon.
.SH SYNOPSIS
momctl [\-c {jobid|all}] [\-C] [\-d {integer|jobid}] [-f file] [\-H host [,host]...] [\-l] [-p port] [-q attrib] -r {file|LOCAL:file} [-s] [-u]
.SH DESCRIPTION
The 
.B momctl
command allows remote shutdown, reconfiguration, diagnostics, and querying of the pbs_mom daemon.
.NL
.SH OPTIONS
.IP "\-c {jobid|all}" 15
Make the MOM unaware of the job's existence. It does not clean up any processes associated with the job.
.IP "\-C" 15
Cycle pbs_mom (force the MOM to send a status update to pbs_server).
.NL
.IP "\-d {integer|jobid}" 15
Diagnose MOM(s). (For more details, see DIAGNOSE DETAIL below.)
.NL
.IP "\-f file" 15
A file containing only comma or whitespace (space, tab, or new line) delimited hostnames.
.NL
.IP "\-h host[,host]..." 15
A comma separated list of hosts.
.NL
.IP "\-l" 15
Display the layout of the machine as it is understood by Torque.
.NL
.IP "\-p port" 15
The port number for the specified MOM(s).
.NL
.IP "\-q attrib" 15
Query or set attrib, where attrib is a property listed by pbsnodes -a (see QUERY ATTRIBUTES below for a list of attributes). Can also be used to query or set pbs_mom options (see RESOURCES below).
.NL
.IP "\-r {file|LOCAL:file}" 15
Reconfigure MOM(s) with remote or local config file, file. This does not work if $remote_reconfig is not set to true when the MOM is started.
.NL
.IP "\-s" 15
Have the MOM shut itself down gracefully (this includes reporting to server so that pbsnodes marks the node down).
.NL
.IP "\-u"
Update the hardware configuration on pbs_server with a layout from the MOM.
.NL
.SH QUERY ATTRIBUTES
.IP "\fBarch\fP" 4
node hardware architecture
.IP "\fBavailmem\fP" 4
available RAM
.IP "\fBloadave\fP" 4
1 minute load average
.IP "\fBncpus\fP" 4
number of CPUs available on the system
.IP "\fBnetload\fP" 4
total number of bytes transferred over all network interfaces
.IP "\fBnsessions\fP" 4
number of sessions active
.IP "\fBnusers\fP" 4
number of users active
.IP "\fBphysmem\fP" 4
configured RAM
.IP "\fBsessions\fP" 4
list of active sessions
.IP "\fBtotmem\fP" 4
configured RAM plus configured swap
.NL
.SH RESOURCES
Resource Manager queries can be made with momctl -q options to retrieve and set pbs_mom options. Any configured static resource may be retrieved with a request of the same name. These are resource requests not otherwise documented in the PBS ERS.
.br
.IP "\fBcycle\fP" 4
Forces an immediate MOM cycle.
.IP "\fBstatus_update_time\fP" 4
Retrieve or set the $status_update_time parameter.
.IP "\fBcheck_poll_time\fP" 4
Retrieve or set the $check_poll_time parameter.
.IP "\fBconfigversion\fP" 4
Retrieve the config version.
.IP "\fBjobstartblocktime\fP" 4.IP "\fB
Retrieve or set the $jobstartblocktime parameter.
.IP "\fBenablemomrestart\fP" 4
Retrieve or set the $enablemomrestart parameter.
.IP "\fBloglevel\fP" 4
Retrieve or set the $loglevel parameter.
.IP "\fBdown_on_error\fP" 4
Retrieve or set the $down_on_error parameter.
.IP "\fBdiag0 - diag4\fP" 4
Retrieves varied diagnostic information.
.IP "\fBrcpcmd\fP" 4
Retrieve or set the $rcpcmd parameter.
.IP "\fBversion\fP" 4
Retrieves the pbs_mom version.
.NL
.SH DIAGNOSE DETAIL
.IP "\fBLevel 0\fP" 4
Display the following information: Local hostname, Expected server hostname, Execution version, MOM home directory. MOM config file version (if specified), Duration MOM has been executing, Duration since last request from pbs_server daemon, Duration since last request to pbs_server daemon, RM failure messages (if any), Log verbosity level, Local job list
.IP "\fBLevel 1\fP" 4
All information for level 0 plus the following: Interval between updates sent to server, Number of initialization messages sent to pbs_server daemon, Number of initialization messages received from pbs_server daemon, Prolog/epilog alarm time, List of trusted clients
.IP "\fBLevel 2\fP" 4
All information from level 1 plus the following: PID, Event alarm status
.IP "\fBLevel 3\fP" 4
All information from level 2 plus the following: syslog enabled
.NL
.RE



CHANGELOG
c - crash     b - bug fix    e - enhancement    f - new feature  n - note

NOTE: the CHANGELOG file is now deprecated. Please check the release notes page
on the Adaptive Computing Website. For example, the 6.0.1 release notes can be
found:

http://docs.adaptivecomputing.com/9-0-1/releaseNotes/help.htm

6.0.0
  b - TRQ-3245. Enable reporter mom to correctly handle UNKNOWN role.
  b - TRQ-3242. Fix problem where resource string argument to prologue
      script getting garbled.
  b - TRQ-3232 Start threadpool at pbs_mom start.
  b - TRQ-3117. Fix a misspelling in the number_successful tag (was number_successfull)
  e - TRQ-3131. Add capability to pass environment variables to pbsdsh. 
  b - TRQ-3185. Create subdirs when server attribute use_job_subdirs set.


5.1.2
  b - TRQ-2675. Fix small errors in suse init.d scripts.
  b - TRQ-3235. Fix problem when path to error, output or execution environment
      contains one or more spaces.
  e - TRQ-3098. Add the ability to set a parameter exit_code_canceled_job to force all
      canceled jobs to have the same exit code regardless of the state they were in
      when they were canceled.
  e - TRQ-2836. Make node health check run on sister nodes when configured for job
      start and job end as well.
  e - TRQ-2843. Add the qmgr setting dont_write_nodes_file to make it so that nodes
      cannot be edited dynamically
  f - TRQ-2897. Add the ability to adopt running processes into a job with pbs_track.
  b - TRQ-3189. Never delete a running job because of a dependency.

5.1.1.2
  e - TRQ-3197. Add support for RHEL7 and SLES12.

5.1.1
  b - TRQ-2947. Fix a race condition on deleting jobs which are failing to start.
  c - TRQ-3068. Fix a race condition where a job may be deleted but have it's pointer
      may still be in the alljobs container.
  b - TRQ-2753. Fix a memory leak in generating the authoritative okclients list.
  b - TRQ-2332. Fix a job dependency problem when the failover server comes up. This
      only affects users running high availability.
  b - TRQ-3023. Fix a bug when ALPS incorrectly returns a permanent confirmation failure.
  b - TRQ-2833. Set CUDA_VISIBLE_DEVICES to only the indices for this host when
      it will be set. 
  b - TRQ-3039. Fix a deadlock when deleting a job where other jobs have after any dependencies
      on the first job.
  f - TRQ-2782. Distribute job files into subdirectories when server attribute use_jobs_subdirs
      set to true. Default is false (do not distribute job files).
  b - TRQ-3116. Make qsub only retry on transient errors.
  b - TRQ-3122. Fix a problem with login_property not working correctly (cray only).
  b - TRQ-3114. Fix an issue where an asynchronously started job is stuck with a 
      substate of starting after a failed job start.
  b - TRQ-3110. Handle slot limits correctly when jobs are preempted.
  e - TRQ-3095. Add the server setting disable_automatic_requeue to stop jobs from being
      requeued if they experience a transient failure on the mom.
  e - TRQ-2307. Fix probelms where mom restarts intermittently fail.
  b - TRQ-2946. Make qmgr able to handle Cray numeric node ids.
  b - TRQ-2790. Make offlining cray compute nodes persist across restarts. 
  e - TRQ-3104. Add millisecond precision to the Torque log file
  e - TRQ-2881. Add node health check error messages to a node's notes and therefore pbsnodes
      output.
  b - TRQ-3166. Add another safety check before killing stray jobs.


5.0.2
  b - TRQ-3029. Make it so that pbs_server can't have active threads when the main
      thread exits.
  b - TRQ-3012. Fix memory leaks that happen each time a job is run.
  b - TRQ-2966. Improved job rerun speed which had been significantly slowed down
      starting in 4.2.6. Also, pbs_mom now correctly accounts for job resources when
      user jobs call setsid more than once.
  c - TRQ-2987. Fix a crash around job exits due to incorrect error code handling.
  b - TRQ-2841. Fix some ways that max_user_queuable can become incorrect
  b - TRQ-3097. Fixed a problem where failed job submissions would count against
      the max_user_queuable count and could not be cleared until pbs_server
      was restarted.
  b - TRQ-3087. Fixed a problem where completed jobs were counted against max_user_queuable
      when restarting pbs_server. Also if the max_user_queuable was set on a
      queue and the number of queued jobs and completed jobs were over the
      maximum then the last jobs submitted would not get loaded.
 

5.0.1.h2
  b - Reverted a change in 5.0.0 which made it so a user could not submit a
      job from a node which had been allowed using the acl_hosts list. The
      change to 5.0.0 made it so the ruserok call could not be made to check
      for user authorization.

5.0.1
  e - TRQ-2410. Improved qstat behavior in cases where bad job IDs were referenced
      in the command.
  e - TRQ-2460. Two new fields were added to the accounting file for completed jobs:
      total_execution_slots and unique_node_count. total_execution_slots should be
      20 for a job that requests nodes=2:ppn=10. unique_node_count should be the
      number of unique hosts the job occupied.
  e - TRQ-2594. TORQUE now uses the Munge API rather than forking when configured
      with the --enable-munge-auth option.
  e - TRQ-2863. Reduced verbosity in error logging in HA environments.
  e - TRQ-2868. TORQUE now allows for the modification of the output location
      based on the Mother superior hostname. An environment variable ($HOSTNAME)
      has been added to the job's environment.
  e - TRQ-2882. Improved trqauthd error messages to more meaningful and less
      redundant.
  e - TRQ-2890. Added stderr capturing when using -o option.

src/momctl/momctl.c
/* momctl */

/* build w/

  cc momctl.c -o momctl -L ../lib/Libnet -L ../lib/Libpbs -lnet -lpbs -I ../include

 */




#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include <errno.h>

extern char *optarg;

#include "net_connect.h"
#include "dis.h"
#include "mcom.h"

#include "pbs_error.h"
#include "pbs_ifl.h"
#include "resmon.h"
#include "rm.h"
#include "log.h"
#include "../lib/Liblog/chk_file_sec.h"
#include "lib_ifl.h"

#define MAX_QUERY  128

const char *LocalHost = "localhost";

const char *DiagPtr  = "diag";

char *Query[MAX_QUERY];
int   QueryI   = 0;

char *FPtr     = NULL;
char *JPtr     = NULL;

char HostFile[256];
char ConfigBuf[65536];

int MOMPort = 0;  /* use default PBS MOM port */

mbool_t IsVerbose = FALSE;


enum MOMCmdEnum
  {
  momNONE = 0,
  momClear,
  momQuery,
  momReconfig,
  momShutdown,
  momLayout
  };

enum MOMCmdEnum CmdIndex = momNONE;


/* prototypes */

void MCShowUsage(const char *);
int do_mom(char *, int, int);

/* END prototypes */




int perform_communications_with_retry(

  char *hostname,
  int   MOMPort,
  int  *fail_count)

  {
  int retries = 0;
  int local_errno;
  int rc = PBSE_NONE;

  while (retries < 5)
    {
    rc = do_mom(hostname, MOMPort, CmdIndex);
    if (rc >= 0)
      break;

    *fail_count = *fail_count + 1;
    
    local_errno = -1 * rc;

    switch (local_errno)
      {
      case EACCES:
      case EPERM:
      case ECONNREFUSED:
      case ENETUNREACH:
      case EFAULT:
      case EAFNOSUPPORT:

        return(rc);

        break;
      }
    
    fprintf(stdout, "attempting command again\n");
    retries++;
    sleep(1);
    continue;
    }

  return(rc);
  } /* END perform_communications_with_retry() */




int main(

  int    ArgC,  /* I */
  char **ArgV)  /* I */

  {
  const char *OptString = "c:Cd:f:h:lp:q:r:sv";

  char  HostList[65536];

  char *HPtr;


  int c;
  int rc = PBSE_NONE;
  int local_errno = 0;

  int HostCount;
  int FailCount;

  /* initialize */

  HostList[0]     = '\0';
  ConfigBuf[0] = '\0';

  if (IamRoot() == 0)
    {
    exit(EXIT_FAILURE);
    }

  while ((c = getopt(ArgC, ArgV, OptString)) != EOF)
    {
    switch (c)
      {

      case 'c':

        /* clear job */

        JPtr = optarg;

        CmdIndex = momClear;

        break;

      case 'C':

        /* force cycle */

        CmdIndex = momQuery;

        Query[QueryI] = strdup("cycle");

        QueryI++;

        break;

      case 'd':

        /* diagnose */
        /* FORMAT:  momctl -d<X> */

        CmdIndex = momQuery;

        if ((Query[QueryI] = (char *)calloc(strlen(DiagPtr) + 3, sizeof(char))) == NULL)
      	  {
          fprintf(stderr,"ERROR:    could not calloc %d bytes!\n",
            (int)strlen(DiagPtr) + 3);

          exit(EXIT_FAILURE);
      	  }

        if (optarg == NULL)
          {
          strncpy(Query[QueryI],DiagPtr,strlen(DiagPtr));
          }
        else
          {
          snprintf(Query[QueryI],strlen(DiagPtr) + 2,"%s%s",
            DiagPtr,
            optarg);
          }

        QueryI++;

        break;

      case 'f':

        {
        int   rc;

        FILE *fp;

        long   size;

        if ((fp = fopen(optarg, "r")) == NULL)
          {
          fprintf(stderr, "ERROR:    cannot open file '%s', errno: %d (%s)\n",
            optarg,
            errno,
            strerror(errno));

          exit(EXIT_FAILURE);
          }

        rc = fread(HostList, sizeof(HostList), 1, fp);

        if ((rc == 0) && (!feof(fp)))
          {
          fprintf(stderr, "ERROR:    cannot read file '%s', errno: %d (%s)\n",
            optarg,
            errno,
            strerror(errno));

          exit(EXIT_FAILURE);
          }

        if ((size = ftell(fp)) < 0)
          size = 0;

        HostList[MIN(size,(long)sizeof(HostList) - 1)] = '\0';

        fclose(fp);
        }  /* END BLOCK */

      break;

      case 'h':

        /* connect to specified host */

        snprintf(HostList, sizeof(HostList), "%s", optarg);

        break;

      case 'l':

        CmdIndex = momLayout;

        break;

      case 'p':

        /* port */

        if (optarg == NULL)
          MCShowUsage("port not specified");

        MOMPort = (int)strtol(optarg, NULL, 10);

        if (MOMPort == 0)
          MCShowUsage("invalid port specified");

        break;

      case 'q':

        /* query resources */

        if (optarg == NULL)
          {
          MCShowUsage("query not specified");

          Query[QueryI] = strdup(DiagPtr);
          }
        else
          {
          Query[QueryI] = strdup(optarg);
          }

        QueryI++;

        CmdIndex = momQuery;

        break;

      case 'r':

        /* reconfigure */

        {
        CmdIndex = momReconfig;

        /* NOTE:  specify remote file to load -> 'fname' */
        /*        specify local file to stage -> 'LOCAL:fname' */

        if (optarg == NULL)
          MCShowUsage("file not specified");

        if (!strncmp(optarg, "LOCAL:", strlen("LOCAL:")))
          {
          FILE *fp;

          int   size;

          int   rc;

          char *ptr;
          char *cptr;

          strcpy(ConfigBuf, "CONFIG:");

          cptr = ConfigBuf + strlen(ConfigBuf);

          ptr = optarg + strlen("LOCAL:");

          if ((fp = fopen(ptr, "r")) == NULL)
            {
            fprintf(stderr, "ERROR:    cannot open file '%s', errno: %d (%s)\n",
                    optarg,
                    errno,
                    strerror(errno));

            exit(EXIT_FAILURE);
            }

          rc = fread(cptr, sizeof(ConfigBuf) - strlen(ConfigBuf), 1, fp);

          if ((rc == 0) && (!feof(fp)))
            {
            fprintf(stderr, "ERROR:    cannot read file '%s', errno: %d (%s)\n",
                    optarg,
                    errno,
                    strerror(errno));

            exit(EXIT_FAILURE);
            }

          size = ftell(fp);

          ConfigBuf[MIN(size + strlen("CONFIG:"),sizeof(ConfigBuf) - 1)] = '\0';

          fclose(fp);
          }
        else
          {
          snprintf(ConfigBuf, sizeof(ConfigBuf), "%s", optarg);
          }
        }  /* END (case 'r') */

      break;

      case 's':

        /* shutdown */

        CmdIndex = momShutdown;

        break;

      case 'v':

        /* report verbose logging */

        IsVerbose = TRUE;

        break;
      }  /* END switch (c) */
    }    /* END while (c = getopt()) */

  if (CmdIndex == momNONE)
    {
    MCShowUsage("no command specified");
    }

  if (HostList[0] == '\0')
    snprintf(HostList, sizeof(HostList), "%s", LocalHost);

  HPtr = strtok(HostList, ", \t\n");

  HostCount = 0;

  FailCount = 0;

  /* at this point, all args processing and setup is completed ...
   * ... now we run through each comma-delimited word in HPtr */

  while (HPtr != NULL)
    {
    if ((*HPtr == ':') && (*(HPtr + 1) != '\0'))
      {
      /* finds nodes with this property */

      int con;
      char *def_server, *pserver, *servername;

      struct batch_status *bstatus, *pbstat;

      struct attrl *nodeattrs;

      def_server = pbs_default();

      if ((pserver = strchr(HPtr,'@')) != NULL)
        {
        *pserver = '\0';
        servername = pserver + 1;
        }
      else
        {
        servername = def_server;
        }

      con = pbs_connect(servername);

      if (con < 0)
        {
        fprintf(stderr,"failed to connect to pbs_server:%s\n",
          servername);

        exit(EXIT_FAILURE);
        }

      /* get a batch_status entry for each node in ":property" */

      bstatus = pbs_statnode_err(con,HPtr,NULL,NULL, &local_errno);

      if (bstatus != NULL)
        {
        for (pbstat = bstatus;pbstat != NULL;pbstat = pbstat->next)
          {
          /* check state first, only do_mom() if not down */

          for (nodeattrs = pbstat->attribs;nodeattrs != NULL; nodeattrs = nodeattrs->next)
            {
            if (!strcmp(nodeattrs->name, ATTR_NODE_state))
              {
              if (!strstr(nodeattrs->value, ND_down))
                {
                if ((rc = perform_communications_with_retry(pbstat->name, MOMPort, &FailCount)) == PBSE_NONE)
                  HostCount++;
                }
              else
                {
                fprintf(stderr,"%12s:   skipping down node\n",
                  pbstat->name);
                }

              break;
              }  /* END if (attrib name eq state) */
            }    /* END for (nodeattrs) */
          }      /* END for (pbstat) */

        pbs_statfree(bstatus);
        }  /* END if (bstatus != NULL) */
      else
        {
        fprintf(stderr,"no nodes found in %s on %s\n",
          HPtr,
          servername);
        }

      pbs_disconnect(con);

      if (pserver != NULL)
        *pserver = '@';
      }
    else
      {
      if ((rc = perform_communications_with_retry(HPtr, MOMPort, &FailCount)) == PBSE_NONE)
        HostCount++;
      } /* END if (*HPtr == ':') */

    HPtr = strtok(NULL, ", \t\n");
    }  /* END while (HPtr != NULL) */


  if (IsVerbose == TRUE)
    {
    fprintf(stdout, "Node Summary:  %d Successful  %d Failed\n",
            HostCount,
            FailCount);
    }

  /* test success of do_mom before returning success */
  if (rc != PBSE_NONE)
    exit(EXIT_FAILURE);
    
  /* SUCCESS */

  exit(EXIT_SUCCESS);
  }  /* END main() */




int send_command(

  struct tcp_chan *chan,
  int              cmd)

  {
  int rc;

  if((rc = diswsi(chan,cmd)) != DIS_SUCCESS)
    return (rc);

  if (cmd == RM_CMD_CONFIG)
    {
    if((rc = diswst(chan,ConfigBuf)) != DIS_SUCCESS)
      return(rc);
    }

  DIS_tcp_wflush(chan);
  if (cmd == RM_CMD_CLOSE)
    DIS_tcp_close(chan);

  return(rc);
  } /* END send_command() */





/*
 * send_command_str
 *
 * @param stream - the stream we're writing to
 * @param cmd - the command we're writing
 */

int send_command_str(

  struct tcp_chan *chan,
  int              cmd,
  char            *query) /* I */

  {
  int rc;
  
  rc = diswsi(chan, cmd);

  if ((rc != DIS_SUCCESS) ||
      (cmd == RM_CMD_CLOSE))
    return(rc);

  if (cmd == RM_CMD_CONFIG)
    {
    if((rc = diswst(chan, ConfigBuf)) != DIS_SUCCESS)
      return (rc);
    }

  rc = diswcs(chan, query, strlen(query));

  DIS_tcp_wflush(chan);

  return(rc);
  } /* END send_command_str() */





int check_success(

  struct tcp_chan *chan)

  {
  int rc;
  int local_errno;
  int status = disrsi(chan,&rc);

  if (rc != DIS_SUCCESS)
    return(rc);
  else if (status != RM_RSP_OK)
    {
#ifdef ENOMSG
    local_errno = ENOMSG;
#else
    local_errno = EINVAL;
#endif

    return(-1 * local_errno);
    }
  else
    return(PBSE_NONE);
  } /* END check_success() */




char *read_mom_reply(

  int             *local_errno, /* O */
  struct tcp_chan *chan)

  {
  int   rc;
  char *value = NULL;

  if (check_success(chan))
    return(NULL);

  value = disrst(chan,&rc);

  if (rc != DIS_SUCCESS)
    {
    if (value != NULL)
      free(value);

    *local_errno = rc;

    return(NULL);
    }

  return(value);
  } /* END read_mom_reply() */




int start_dialogue(

  struct tcp_chan *chan)

  {
  int rc;

  rc = diswsi(chan, RM_PROTOCOL);

  if (rc != DIS_SUCCESS)
    return(rc);

  rc = diswsi(chan, RM_PROTOCOL_VER);

  if (rc != DIS_SUCCESS)
    return(rc);

  rc = diswsi(chan, QueryI);

  return(rc);
  } /* END start_dialogue() */




int do_mom(

  char *HPtr,
  int   MOMPort,
  int   CmdIndex)

  {
  int socket;
  int local_errno = 0;
  struct tcp_chan *chan = NULL;
  int rc;

  if ((socket = openrm(HPtr, MOMPort)) < 0)
    {
    /* FAILURE */

    extern char TRMEMsg[];

    fprintf(stderr, "cannot connect to MOM on node '%s', errno=%d (%s)\n",
      HPtr,
      errno,
      strerror(errno));

    if (TRMEMsg[0] != '\0')
      {
      fprintf(stderr, " %s\n",
              TRMEMsg);
      }

    return(socket);
    }
  else if ((chan = DIS_tcp_setup(socket)) == NULL)
    {
    fprintf(stderr, "%s: can not allocate memory of socket buffers\n", __func__);
    return -1;
    }

  /* send protocol and version, plus how many queries we're sending */
  if (QueryI == 0)
    QueryI = 1;

  if (start_dialogue(chan) != DIS_SUCCESS)
    {
    fprintf(stderr,"ERROR:    Unable to write the number of queries to %s (errno=%d-%s)\n",
        HPtr,
        errno,
        strerror(errno));
    
    send_command(chan,RM_CMD_CLOSE);

    DIS_tcp_cleanup(chan);

    return(-1);
    }

  if (IsVerbose == TRUE)
    {
    fprintf(stderr, "INFO:     successfully connected to %s\n",
            HPtr);
    }

  switch (CmdIndex)
    {

    case momClear:

      {
      char tmpLine[1024];

      char *Value;

      snprintf(tmpLine, 1024, "clearjob=%s",
               (JPtr != NULL) ? JPtr : "all");

      if (send_command_str(chan, RM_CMD_REQUEST, tmpLine) != 0)
        {
        /* FAILURE */

        fprintf(stderr,"ERROR:    cannot request job clear on %s (errno=%d-%s)\n",
          HPtr,
          errno,
          strerror(errno));

        send_command(chan,RM_CMD_CLOSE);

        return(-1);
        }

      if ((Value = (char *)read_mom_reply(&local_errno, chan)) == NULL)
        {
        /* FAILURE */

        fprintf(stderr,"ERROR:    job clear failed on %s (errno=%d - %s: %d - %s)\n",
          HPtr,
          errno,
          pbs_strerror(errno),
          local_errno,
          pbs_strerror(local_errno));

        send_command(chan,RM_CMD_CLOSE);

        return(-1);
        }

      /* job cleared */

      fprintf(stdout,"job clear request successful on %s\n",
        HPtr);

      free(Value);
      }  /* END BLOCK (case momClear) */

    break;

    case momShutdown:

      {
      if ((send_command(chan,RM_CMD_SHUTDOWN) != PBSE_NONE) ||
          (check_success(chan) != PBSE_NONE))
        {
        /* FAILURE */

        fprintf(stderr,"ERROR:    cannot shutdown mom daemon on %s (errno=%d-%s)\n",
          HPtr,
          errno,
          pbs_strerror(errno));

        send_command(chan,RM_CMD_CLOSE);

        exit(EXIT_FAILURE);
        }

      fprintf(stdout, "shutdown request successful on %s\n",
        HPtr);
      }    /* END BLOCK */

    break;

    case momReconfig:

      {
      if ((send_command(chan,RM_CMD_CONFIG) != PBSE_NONE) ||
          (check_success(chan) != PBSE_NONE))
        {
        /* FAILURE */

        fprintf(stderr,"ERROR:    cannot reconfigure mom on %s (errno=%d-%s)\n",
          HPtr,
          errno,
          pbs_strerror(errno));

        send_command(chan,RM_CMD_CLOSE);

        return(-1);
        }

      fprintf(stdout, "reconfig successful on %s\n",
        HPtr);
      }  /* END BLOCK (case momReconfig) */

    break;

    case momLayout:
        
      char *value;

      if (send_command(chan, RM_CMD_LAYOUT) != PBSE_NONE)
        {
        fprintf(stdout, "Layout command failed to send to mom\n");
        return(-1);
        }
        
      if ((value = read_mom_reply(&local_errno, chan)) == NULL)
        {
        fprintf(stdout, "Could not read a layout reply from the mom\n");
        return(-1);
        }
      else
        {
        fprintf(stdout, "%s", value);
        free(value);
        }

      break;

    case momQuery:

    default:

      {
      char *ptr;

      int  rindex;

      char *Value;

      for (rindex = 0; rindex < QueryI; rindex++)
        {
        if (send_command_str(chan, RM_CMD_REQUEST, Query[rindex]) != 0)
          {
          fprintf(stderr,"ERROR:    cannot add query for '%s' on %s (errno=%d-%s)\n",
            Query[rindex],
            HPtr,
            errno,
            pbs_strerror(errno));
          }
        }

      for (rindex = 0;rindex < QueryI;rindex++)
        {
        if ((ptr = strchr(Query[rindex],'=')) != NULL)
          {
          *ptr = '\0';
          }

        if ((Value = (char *)read_mom_reply(&local_errno, chan)) == NULL)
          {
          fprintf(stderr, "ERROR:    query[%d] '%s' failed on %s (errno=%d - %s : %d - %s)\n",
            rindex,
            Query[rindex],
            HPtr,
            errno,
            pbs_strerror(errno),
            local_errno,
            pbs_strerror(local_errno));
            return(-1);

          }
        else
          {
          if (!strncmp(Query[rindex], "diag", strlen("diag")))
            {
            fprintf(stdout, "%s\n",
              Value);
            }
          else if (!strncmp(Query[rindex], "cycle", strlen("cycle")))
            {
            fprintf(stdout, "mom %s successfully cycled %s\n",
              HPtr,
              Value);
            }
          else
            {
            fprintf(stdout, "%12s: %12s = '%s'\n",
              HPtr,
              Query[rindex],
              Value);
            }
          }

        free(Value);

        if (ptr != NULL)
          {
          *ptr = '=';
          }
        }  /* END for (rindex) */
      }    /* END BLOCK (case momQuery) */

    break;
    }  /* END switch(CmdIndex) */
  rc = diswsi(chan, RM_PROTOCOL);

  if (rc != DIS_SUCCESS)
    goto do_mom_fail;

  rc = diswsi(chan, RM_PROTOCOL_VER);

  if (rc != DIS_SUCCESS)
    goto do_mom_fail;

  rc = diswsi(chan, 1);

  if (rc != DIS_SUCCESS)
    goto do_mom_fail;

  /* send_command will free chan */
  send_command(chan,RM_CMD_CLOSE);

  return(0);

do_mom_fail:
  DIS_tcp_close(chan);
  return(rc);

  } /* END do_mom() */







void MCShowUsage(

  const char *Msg)  /* I (optional) */

  {
  if (Msg != NULL)
    fprintf(stderr, "  %s\n",
      Msg);

  fprintf(stderr, "USAGE:  momctl <ARGS>\n");

  fprintf(stderr, "            [ -c {JOB|'all'} ]    // CLEAR JOB\n");

  fprintf(stderr, "            [ -C ]                // CYCLE\n");

  fprintf(stderr, "            [ -d DIAGLEVEL ]      // DIAGNOSE (0 - 3)\n");

  fprintf(stderr, "            [ -f HOSTFILE ]       // FILE CONTAINING HOSTLIST\n");

  fprintf(stderr, "            [ -h HOST[,HOST]... ] // HOSTLIST\n");

  fprintf(stderr, "            [ -l ]                // PRINT MACHINE LAYOUT (CGROUPS ONLY)\n");

  fprintf(stderr, "            [ -p PORT ]           // PORT\n");

  fprintf(stderr, "            [ -q ATTR ]           // QUERY ATTRIBUTE\n");

  fprintf(stderr, "            [ -r FILE ]           // RECONFIG\n");

  fprintf(stderr, "            [ -s ]                // SHUTDOWN\n");

  fprintf(stderr, "\n");

  fprintf(stderr, " Only one of c, C, d, q, r, or s must be specified, but -q may\n");

  fprintf(stderr, " be used multiple times. HOST may be a hostname or \":property\".\n");

  exit(EXIT_FAILURE);
  }  /* END MCShowUsage() */



/* END momctl.c */

